connector/connector.go

```diff
@@ -41,6 +44,12 @@ type Traces interface {
 	consumer.Traces
 }
 
+// TracesRouter feeds the first consumer.Traces in each of the specified pipelines.
+type TracesRouter interface {
```

Comment:
There are few places where we do fan-out:
* Receivers
* Connector as receiver
* Last processor

We need to clarify in which case we expect this interface to work (what is the expected goal).

Reply:
My goal was only with connectors, so I defined the interface in the connectors package. We could allow it in the other cases as well, but I think any use case that would require it could be implemented as a connector. In my opinion, this should remain a connectors-only interface unless there is a clear need.

Reply:
"A last processor" use-case may also be ok. It simplifies a bit the configuration, for use-cases like the routing processor which can remain a processor.

If we don't allow that, what about we disallow this by doing some wrapping in case of processors/receivers to not be able to convert the next to this interface.

Reply:
> "A last processor" use-case may also be ok. It simplifies a bit the configuration, for use-cases like the routing processor which can remain a processor. Not proposing, want to hear ideas. I also believe that it is clear that all "routing" happens in the connectors.

I think there is a better case to do routing as a connector instead of a processor. For example, with logs it is sometimes necessary to read from one file but then separate based on the structure. Then process each type differently based on that. A connector can support this easily but a processor would have to send to an exporter that _is_ a connector, so I think it's easier to have only connectors behave this way.

Similarly with metrics, maybe you want to process metrics differently for some reason - name, type, attributes - so you send to different pipelines for differentiated processing. A processor can only send to different exporters, so you would have to sent to exporters that are connectors anyways.

> If we don't allow that, what about we disallow this by doing some wrapping in case of processors/receivers to not be able to convert the next to this interface.

I agree. This is what I have intended with this implementation. The service only passes the `connector.TracesRouter` type to create connector methods, so there is no worry about this for other components. Even if we need to change the signature for creating connectors, we would still make sure in the service to only pass in the routing types to connectors.

Reply:
One more reason to only allow routing from connectors is that accessing a fanout is only possible for the last processor in a pipeline. Users would have to take this into account in order to determine valid placement of the processor. Connectors are always in the correct position to handle this, so the only consideration for users is which pipelines they are emitting to, which is a standard concern for connectors.